### PR TITLE
Coordinator model

### PR DESCRIPTION
**Pull request by Amstutz**
Dear ILIAS Developers,

The TB  proposes to introduce a new maintenance model as an alternative to the classic component maintainer model suited for certain very specific scenarios (see proposed doc). We intend to discuss this during the next JF. You are very welcome to give feedback or ask for changes. Thx a lot in advance.

### CODE REVIEW

**Review by chfsx — APPROVED** [review]

**Comment by colin-kiegel** [comment]
[Rendered Version](https://github.com/Amstutz/ILIAS/blob/coordinator_model/docs/documentation/maintenance-coordinator.md)

File: docs/documentation/maintenance-coordinator.md
Comment (colin-kiegel) [inline docs/documentation/maintenance-coordinator.md]: @Amstutz github doesn't render this as headline without a space (see also other headlines below)

**Review by colin-kiegel — CHANGES_REQUESTED** [review]
Sounds good to me.

- Could you add a sentence about "who defines the Coordinator"?
- Did any concerns arise in the TB discussion? If so, can you give a brief summary?
- Can you link the definition of a contributor somewhere?
- what happens if a former contributor doesn't maintain his contribution anymore (for whatever reason, different job, etc.)?

File: docs/documentation/maintenance-coordinator.md
Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md]: Thx, fixed.

Comment (colin-kiegel) [reply docs/documentation/maintenance-coordinator.md]: not very important, but this also affects the last four headlines:
- Change Management
- Issue Management
- Scenarios
- What can be expected of a coordinator?

Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md]: Thx again ;-)

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [inline docs/documentation/maintenance-coordinator.md:75]: I like this part, but this part needs more definition.
How is this vision defined? Do the coordinators have to create Roadmaps(format, platform etc.)? Should the coordinator create READMEs what contributions should be made?

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [inline docs/documentation/maintenance-coordinator.md:79]: This is a tough one.
Considering following scenario:
* A contribution comes in
* The contributer doesn't want this changes, because the changes are to big and can not be reviewed in pleasing amount of time or the side effects can't be calculated.
* The JF wants these changes.
* The TB agrees with the JF to merge these changes.

Who maintains the code and documentation?
Who is responsible for upcoming bugs the contributer, maintainer, the technical board or the Jour Fixe?

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [inline docs/documentation/maintenance-coordinator.md:92]: What happens if there is no responsible developer anymore(other job, not feeling responsible, death)? What alternatives are there?

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [inline docs/documentation/maintenance-coordinator.md:6]: Wouldn't it be good to add chapter of becoming a coordinator? I'm currently missing this part, but maybe somone can enlighten me :+1:

Also who decides demote the person from being a coordinator?

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [inline docs/documentation/maintenance-coordinator.md:17]: Why not change this name to curator then? :sweat_smile:

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [inline docs/documentation/maintenance-coordinator.md:89]: What about CI tools to make it easier for the coordinator to run the tests? Otherwise the coordinator needs the check out **every single PR** and execute the tests.

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [inline docs/documentation/maintenance-coordinator.md:61]: How will the focus on code quality be assured? Will the branches set as [protected](https://help.github.com/articles/about-protected-branches/)?

File: docs/documentation/maintenance-coordinator.md
Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md:17]: Good point indeed this was in discussion, but we decided against, since we believe it hard to understand (curator might not be such a commonly known word). However, there have been voices prefering "curator". What would you prefer?

File: docs/documentation/maintenance-coordinator.md
Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md:6]: Hold up your hand if one is looked for ;-). Currently we do not expect struggling with to many willing coordinators. However you are right. We should lose a word or two over this. In my understanding, the PM would make the final decision in case of some argument. Note that is my personal understanding of this matter, we did not discuss this in the TB.

File: docs/documentation/maintenance-coordinator.md
Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md:92]: Good point. This is related to the problem in the component maintainer model. What if a maintainer leaves without any body willing to fill it's shoes? Also, note that this is a problem for currently unmaintained parts of the code (parts folders without anybody entered in the maintainance.json). If an issue is not assignable to anyone it remains on the coordinators desk.  Further steps are up to the coordinator (look for funding, communicate the issue to the other devs and check for possible replacement).

File: docs/documentation/maintenance-coordinator.md
Comment (klees) [reply docs/documentation/maintenance-coordinator.md:61]: In the end there is no assurance, we somehow need to trust the coordinator that he pursues that goal and treats contributions accordingly. Technical measures won't be able to guarantee anything. This still makes the question about demotion more interesting, though. Also I'm not against protected branches or any technical measures that makes coordinators, contributors or anyones lives easier, but to me this is not about guarantees.

Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md:61]: Through detailed discussion of the interfaces beforehand and through review of contributions through code quality. Since multiple people will work on the component, we expect the coordinator to introduce further measures such the requirement of unit tests.

File: docs/documentation/maintenance-coordinator.md
Comment (klees) [reply docs/documentation/maintenance-coordinator.md:75]: IMO the specifics will heavily depend on the component we are talking about and also on the state of the component, so beeing vague is fine for me here. This might be a point we want to elaborate on in the future.

File: docs/documentation/maintenance-coordinator.md
Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md:79]: The contributer. The coordinator will most certainly assign bugs directly to him/her.

But it's worth looking closer at the given scenario. We believe it to be rare, that the JF would very much disagree with the opinion of the coordinator (which is the expert for the component and has good reasons for or against something). After this a majority of the TB members would need to agree, that the coordinator is wrong and this component should be added. This could only happen if there are strong reasons for it.

This somehow also answers the question, of a coordinators are replaced. If the coordinator is overruled like this in some crucial decision, a change of coordinator is possible due to the former one maybe stepping down.

Comment (klees) [reply docs/documentation/maintenance-coordinator.md:79]: I think this scenario is only realistic if there are multiple pigheaded or stupid persons involved. I do not expect this rule to be used very soon, especially not if the consequences are dire. This rather is a demand to find a compromise before this rule needs to be used.

File: docs/documentation/maintenance-coordinator.md
Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md:89]: I hope very very much, that the coordinator will use the CI for this (ci.ilias.de). However we still need this nice feature of testing PR's automatically beforehand ;-).

**Comment by Amstutz** [comment]
@colin-kiegel , thanks for your feedback.

> Could you add a sentence about "who defines the Coordinator"?

Do you mean, who decides on who is coordinator for what/who names the coordinator for some component? If yes, see my answer above to @legionth .

> Did any concerns arise in the TB discussion? If so, can you give a brief summary?

This is the result of several iterations of refining an initial proposal. No, we will not summarize this process here. We agree that this is a good basis to discuss this further in a larger group.

> Can you link the definition of a contributor somewhere? what happens if a former contributor doesn't maintain his contribution anymore (for whatever reason, different job, etc.)?

https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-3/docs/documentation/contributing.md. For the other question, see my answer to @legionth .

**Comment by colin-kiegel** [comment]
@Amstutz Thx. I think there was a **misunderstanding** about this question

>> Did any concerns arise in the TB discussion? If so, can you give a brief summary?
> 
> This is the result of several iterations of refining an initial proposal. No, we will not summarize this process here. We agree that this is a good basis to discuss this further in a larger group.

I was not asking about the process or how the TB ended up with this proposal.

My question is:
- Do you, or does the TB _'expect any new challenges'_ with this proposal (.. major ones)?

**Comment by Amstutz** [comment]
@colin-kiegel oke, thx for the clarification. Yes, we expect challanges, however we did not specifically list them, so I can only speak for my self here. I expect it not to be easy to find coordinators for various components, since it can result in quite some work. However we hope, that the perspective to help shaping key parts of ILIAS in a direction that the coordinator believes in, might be a strong incentive to tackle such a task. Another concern has already been pointed out by you and @legionth , what if developers leave?

**Comment by klees** [comment]
@colin-kiegel Problems with bugfixing where also raised as concern in the TB.

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [reply docs/documentation/maintenance-coordinator.md:6]: Maybe we should write it down if every consistent with it. What do you think?

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [reply docs/documentation/maintenance-coordinator.md:17]: Maybe this is just the sentence that bugs me. Describing a role with another role seems a little bit odd to me. But maybe its just me.

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [reply docs/documentation/maintenance-coordinator.md:75]: This is where I would disagree. The problem is that is not clear where and how those visions are defined. The current text implies that these visions can also be in the coordinators head, which would not be compatible with the idea of the coordinator(IMO).

I think the contributers(old and new) and the coordinators of **every component** need centralized contact points to look up what is needed right now. These points should be similar for each component and *could* be defined on GitHub(Issues?), documents in the repository(similar to README) or any external tool.

There is a similar point at the end of this document:
```
The coordinator SHOULD devise some guidelines concerning the processes 
around the respective component fitting it's exact needs as done so for the 
UI-Service. Such guidelines MUST be accepted by the JF.
```

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [reply docs/documentation/maintenance-coordinator.md:89]: Absolutely, thtat is my point here, this would be awesome :+1:. Any reasons why the teamcity hooks are not implemented with this GitHub repo?

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [reply docs/documentation/maintenance-coordinator.md:79]: The problem I have here that this could still happen and the responsibility of the changes is not completely clear. But maybe these are just my concerns, because I've seen similar situations. All I ask is to think about this expectation, that could lead to discomfort between two or more parties.

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [reply docs/documentation/maintenance-coordinator.md:61]: I totally agree with you two :+1: The problem I have is the expectation that could arise with this sentence. The expectation could be that this will get rid of bugs and add code quality just by adding a coordinator.

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [reply docs/documentation/maintenance-coordinator.md:92]: Maybe a sentence would help that the coordinator task is to find appropriate candidates if the candidates are gone. Also a sentence that the coordinator can also address bottlenecks and flaws to the JF or TB to look for help? I don't know if this is common sense in this project, but for newcomers this could be a bit confusing. What do you think?

File: docs/documentation/maintenance-coordinator.md
Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md:75]: Thx. Any concrete proposal we could discuss on the JF?

File: docs/documentation/maintenance-coordinator.md
Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md:92]: I like your reasoning. Any concrete proposal for such a sentence we can discuss in the JF?

File: docs/documentation/maintenance-coordinator.md
Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md:6]: Any concrete proposal for the wording?

File: docs/documentation/maintenance-coordinator.md
Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md:17]: We will ask the JF about the naming, your point is certainly valid.

File: docs/documentation/maintenance-coordinator.md
Comment (Amstutz) [reply docs/documentation/maintenance-coordinator.md:89]: @gvollbach any news on this? What was the reason we do not use those hooks again?

File: docs/documentation/maintenance-coordinator.md
Comment (klees) [reply docs/documentation/maintenance-coordinator.md:75]: @legionth Ok, I definitely see that point. It's about communicating the vision. We already have rules about READMEs in `Services` and `src`. Maybe we could state, that the coordinator should also communicate and document that vision and link to the aforementioned rules.

File: docs/documentation/maintenance-coordinator.md
Comment (klees) [reply docs/documentation/maintenance-coordinator.md:89]: The reason is that anyone could run code on our servers then. Like `exec("rm -rf /")`. We are currently looking into this.

File: docs/documentation/maintenance-coordinator.md
Comment (klees) [reply docs/documentation/maintenance-coordinator.md:61]: This will only happen automatically in a very optimal world... =)

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [reply docs/documentation/maintenance-coordinator.md:6]: I don't know how this procedure should be working, so I can't give you an exact example or an entire chapter. I guess the TB had discussed this how the promotion and demotion should work(Who is deciding? Who is involved etc?)? Would love to hear/read about it :+1:

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [reply docs/documentation/maintenance-coordinator.md:92]: Which would be the default case? Reading your comment the coordinator is encouraged to find a solution by himself/herself without help. If the coordinator can't find a solution in an appropriate time the coordinator must/should escalate the problem to a higher instance (JF or TB? Or both?).

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [reply docs/documentation/maintenance-coordinator.md:89]: @klees is right. I talked with @gvollbach about this. It seems Teamcity acts just like server and doesn't have container-like approach like [Travis](https://travis-ci.org/) or [Jenkins](https://jenkins.io/). I don't want to discuss this here, but I think we all agree that this has to change. I am willing to accept this vague term in this document, but this has to be adjusted when we have operational CI tool.

File: docs/documentation/maintenance-coordinator.md
Comment (legionth) [reply docs/documentation/maintenance-coordinator.md:75]: @Amstutz I don't know what you discussed in the TB. But I would least expect a README with a clear vision or even better a roadmap that will list what needs to be done with assigned developers to each task. If the current tasks are not assignable we could predefined tags like "Needs help" or "Need funding" etc.
@klees This would be great and a good step in the right direction. :+1:.  Any reasons why `Modules` should be left out here?

**Comment by colin-kiegel** [comment]
We discussed the proposal internally at Qualitus - these are our collected feedback and questions

**Feedback**
- We believe the proposal to solve the current major issues with large components.
- We also like the idea to make contributions easier for new developers.
- However we believe the coordinator can be a resource bottleneck. Processes are likely to last longer (to allow reviewing of feature requests and pull requests etc.).
- _The contributor can request funding for his work related to new features._ We suggest to make this a fixed percentage charge on the budget for the new feature, or to introduce some kind of expense allowance by the ILIAS society (.. if there was enough money).

**Questions**
- Who appoints a coordinator?
- What happens if a coordinator changes the company? Does the role remain with the person or company?
- Is it planned to control the coordinator in any way? How?
- Can a coordinator be dropped if things are not going well? Under what conditions? Is there a fixed process?
- We guess that there is _a demarcation of what kind of changes can (a) be made by a new contributor and what kind of changes (b) need to be done by the coordinator or existing contributors._ Is this correct? And is this demarcation formalized in any way or an individual case decision by the coordinator?

**Comment by klees** [comment]
@colin-kiegel and Qualitus:  Thanks **a lot** for this detailed feedback!!

File: docs/documentation/maintenance-coordinator.md
Comment (atoedt) [reply docs/documentation/maintenance-coordinator.md:17]: - Today contributors can prepare Pull Requests. How is this different but in the size of the Pull Request? If it is not different, there is no need for another model. If it is different this should be spelled out. 

- How could this fudge of responsibility with regard to bug fixing be beneficial? I am not convinced that hoping for high quality contributions can be considered a plan. Does a contribution warrant any responsibility and if so which and for how long? 

- How is the incentive issues dealt with i.e. how is ensured that coordinator do not end up with a lot of bug fixing and contributors get all the revenue? Do contributors have to get to the coordinator first and clear their plan or do they just work away and show up with code? 

- I thought the TB is planning for breaking up huge components. Why is this no longer a valid approach? It would keep the responsibility unfidged at least. 

- After reading and re-reading the text, I am still at a loss what the specific expected benefits are and in which specific cases this model should apply. 

- Can coordinators refuse PR that do not comply with their vision / plans for the component point blanc or do they need to substantiate and if so how?

File: docs/documentation/maintenance-coordinator.md
Comment (alex40724) [inline docs/documentation/maintenance-coordinator.md:82]: I think we are missing one sentence here that makes clear, that the coordinator has also the right to reject code changes with good reasons (that need to be stated by the coordinator).

**Review by alex40724 — CHANGES_REQUESTED** [review]

File: docs/documentation/maintenance-coordinator.md
Comment (alex40724) [reply docs/documentation/maintenance-coordinator.md:92]: As a start I suggest to add a sentence like "The coordinator is NOT responsible to fix issues of other developers. Coordinators SHOULD mark unmaintained code and inform the TB about it."

File: docs/documentation/maintenance-coordinator.md
Comment (alex40724) [inline docs/documentation/maintenance-coordinator.md:67]: ...even service providers **(e.g. the UI Framework)**.

File: docs/documentation/maintenance-coordinator.md
Comment (alex40724) [inline docs/documentation/maintenance-coordinator.md:69]: ... parts following a similar scheme **(e.g. the components of the UI Framework)**.

**Comment by klees** [comment]
@colin-kiegel and Qualitus:

We discussed your questions in the TB, these are our answers:

> Who appoints a coordinator?

We propose that the PM and the TB appoint coordinators. We also must confess that we did not explicitely talk about the so far. Thanks for pointing that out =)

> What happens if a coordinator changes the company? Does the role remain with the person or company?

Since the role builds on social capital in the community and a vision of the component it will be near impossible to leave that role at a company. So the role belongs to the person, not the company. On the other hand we expect a coordinator to be replaced more easily than a maintainer, since there should be multiple contributors in the background already.

> Is it planned to control the coordinator in any way? How?

A public control already is exercised by the demand that contributions need to be made via PR and that certain changes need to be announced publicly. This also applies for coordinators themselfs. Also the scenario for which we want to use this model implies that many other developers use the component. Maybe we should point out more clearly how the coordinator himself contributes, so that contributions are always public.

> Can a coordinator be dropped if things are not going well? Under what conditions? Is there a fixed process?

Same as for appointments of coordinators.

> We guess that there is a demarcation of what kind of changes can (a) be made by a new contributor and what kind of changes (b) need to be done by the coordinator or existing contributors. Is this correct? And is this demarcation formalized in any way or an individual case decision by the coordinator?

We expect that this exact line will be established, i.e. new contributors likey won't be allowed to make huge changes immediately. We don't see the requirement to formalize that any further here. Regarding contributions of code, the coordinator is not different from anyone else.

File: docs/documentation/maintenance-coordinator.md
Comment (klees) [reply docs/documentation/maintenance-coordinator.md:17]: @atoedt:

We discussed your questions and remarks in the TB, these are our answers:

> Today contributors can prepare Pull Requests. How is this different but in the size of the Pull Request? If it is not different, there is no need for another model. If it is different this should be spelled out.

The crucial difference is, that the maintainer is a sovereign for his component. He decides what's possible and what's not possible. He may lay down rules, but is not obliged to do so. There are global guidelines he needs to respect, but besides these he has a lot of freedom and room. A contributor completely relies on the courtesy of the maintainer. This is not entirely negative and has advantages (e.g. responsiblity for bugs).

For some components in the software this model is problematic. This is especially the case when it comes to cross-section functionality that everyone has to life with. Coordinators have less freedom and are called upon to lay down rules for others and themselfs, so that contributors can develop an expectation about the component and participate reliably in the development. E.g. is it not possible for @Amstutz and @klees to reject some UI-component without a clear justification. The introduction of UI-components is governed by rules and processes and not the discretion of the two.

In that regard, the coordinator is much less a sovereign, king or owner of a component, but a facilitator for a cooperative development.

**TL/DR:** We define coordinators as a new role and a new maintenance model. Pull request are only an instrument to suggest code changes. We need the new maintenance model for parts of the code where the classic maintainer model does not work well. This need does not depend on the size of pull requests.

> How could this fudge of responsibility with regard to bug fixing be beneficial? I am not convinced that hoping for high quality contributions can be considered a plan. Does a contribution warrant any responsibility and if so which and for how long?

The coordinator gets to decide which code is merged along the rules outline here and in the contribution guideline. He may decide not to merge contributions from people that he deems unreliable or that he suspects won't take part in maintenance of the component. He may also deny to merge code that he deems unreliable or where the risk cannot be assessed. Code will also be contributed in public, i.e. everyone has the possibility to take part in building a reliable component. This is more than some hope for high quality contributions or fudge of responsibility. Even today there is no warrant that a bug gets fixed. However, this should be mentioned clearly in the document to avoid future questions.

Regarding the "maintainer is responsible" principle, the responsibility in the model proposed here is distributed less clearly. On the other hand, the code is supported and maintained by more people, increasing the bus-factor. Also, the maintainer-model can easily collapse, if to much responsibility is borne by a single person.

> How is the incentive issues dealt with i.e. how is ensured that coordinator do not end up with a lot of bug fixing and contributors get all the revenue? Do contributors have to get to the coordinator first and clear their plan or do they just work away and show up with code?

The document already accounts for that question:

* The motivation of the coordinator is mostly driven by the need of a  reliable component for a certain aspect. Further the coordinator is probably the most attractive contractor for clients aiming to change aspects of the component due to the very indepth know how and the listing as coordinator.
* The coordinator MAY ask for funding to perform the review.
* The coordinator moderates the discussion on finding a vision and on the development.

Also see the answer to your bugfixing question.

> I thought the TB is planning for breaking up huge components. Why is this no longer a valid approach? It would keep the responsibility unfidged at least.

There is more than one or the other. The document clearly outlines the scenarios for which we consider this model to be beneficial.

> After reading and re-reading the text, I am still at a loss what the specific expected benefits are and in which specific cases this model should apply.

It is hard to understand that, we tried hard to devise an approachable and comprehensible document and consider the "scenarios"-section to answer your question already. Maybe the problem is that we want to use this model for cross-section functionality. Could it be, that the nature of that functionality is hard to grasp for a non-developer? Maybe we should add examples in the "scenarios"-section (e.g. UI-framework) that further clarify that section.

> Can coordinators refuse PR that do not comply with their vision / plans for the component point blanc or do they need to substantiate and if so how?

The rules from our [contribution guideline](https://github.com/ILIAS-eLearning/ILIAS/blob/release_5-3/docs/documentation/contributing.md) apply here as well. Maybe we should explicitly link them and also rephrase them in the document. I.e.: Yes, contributions can be rejected. The instances to escalate issues are described in the document.

The coordinator is called upon to define additional criteria and processes for the component (e.g. UI-components). This is the gain of this model: PRs cannot be rejected arbitrarily. This allows other developers to build an expectation about the chances of their PR.

**Comment by matthiaskunkel** [comment]
Jour Fixe, 16 JUL 2018 : Pull request and concept is highly apprciated and accepted.

**Comment by Amstutz** [comment]
@klees Do you append changes to the readme? If done, we will merge this.

**Comment by klees** [comment]
@Amstutz Don't know if I will have time to do this before my holidays. I would be happy if someone else took care while I'm away, this will only be merged sometime in august otherwise.

**Comment by Amstutz** [comment]
@klees: Done. Thx for working out the answers.

I tried best to work in the answers of @klees into the doc and I'll merge it. If any body objects any aspect of the doc, please open a new PR and request the change.

**Pull request merged**
